stumpy/core.py

```diff
@@ -2559,3 +2573,183 @@ def _select_P_ABBA_value(P_ABBA, k, custom_func=None):
             MPdist = partition[k]
 
     return MPdist
+
+
+@njit(parallel=True)
+def _merge_topk_PI(PA, PB, IA, IB):
+    """
+    Merge two top-k matrix profiles PA and PB, and update PA (in place) while
+    always choosing values of PA over values of PB in case of ties. Also, update
+    IA accordingly.
+
+    Unlike `_merge_topk_ρI`, where `top-k` largest values are kept, this function
+    keeps `top-k` smallest values.
+
+    `PA` and `PB` are 2D arrays, with each row sorted ascendingly. To update `PA[i]`,
+    the array `PB[i]` is traversed forward from index `0` to its last index, and
+    if its element is smaller than `PA[i, -1]`, i.e. the greatest value in `PA[i]`,
+    then `PA[i]` will be updatd.  In case of tied value `v`, it will be inserted to
+    the right side of the greatest index in `PA[i]` whose value is `v`.
+
+    Parameters
+    ----------
+    PA : numpy.ndarray
+        A (top-k) matrix profile, with ndim of 2, where values in each row are
+        sorted in ascending order.
+
+    PB : numpy.ndarray
+        A (top-k) matrix profile, with ndim of 2, where values in each row are
+        sorted in ascending order. `PB` must have the same shape as `PA`.
+
+    IA : numpy.ndarray
+        A (top-k) matrix profile indices corresponding to PA
+
+    IB : numpy.ndarray
+        A (top-k) matrix profile indices corresponding to PB
+
+    Returns
+    -------
+    None
+    """
+    for i in prange(PB.shape[0]):
+        start = 0
+        stop = np.searchsorted(PA[i], PB[i, -1], side="right")
+
+        if stop == 0:
+            # means `PB[i, -1] < PA[i, 0]`, i.e. the maximum value in `PB[i]` is
+            # less than smallest value in `PA[i]`. So, we should replace `PA[i]`
+            # with `PB[i]` so that we have the top-k smallest.
+            PA[i] = PB[i]
+            IA[i] = IB[i]
+            continue
+
+        for j in range(PB.shape[1]):
+            if PB[i, j] >= PA[i, -1]:
+                # `PB[i]` is sorted ascaendingly.
+                # Hence, in next iteration: `PB[i, j+1] >= PB[i, j] >= PA[i, -1]`
+                break
+
+            # `PB[i, j]` is less than `PA[i, -1]`, the maximum value in `PA[i]`.
+            # so, we must update `PA[i]` to have the top-k smallest values.
+            idx = np.searchsorted(PA[i, start:stop], PB[i, j], side="right") + start
+
+            _shift_insert_at_index(PA[i], idx, PB[i, j], shift="right")
+            _shift_insert_at_index(IA[i], idx, IB[i, j], shift="right")
+
+            start = idx
+            stop += 1  # because of shifting elements to the right by one
+
```

Comment:
Okay, for both `_merge_topk` functions, I think that we've completely over-thought things. I realized that there's an algorithm that is far simpler and at least 2 orders of magnitude faster (without parallelization):

```
@njit
def _merge_topk_PI(PA, PB, IA, IB):
    tmp_P = np.empty(PA.shape[1], dtype=np.float64)
    tmp_I = np.empty(PA.shape[1], dtype=np.int64)
    for i in prange(len(PA)):
        aj, bj = 0, 0
        for k in range(len(tmp_P)):
            if PB[i, bj] < PA[i, aj]:
                tmp_P[k] = PB[i, bj]
                tmp_I[k] = IB[i, bj]
                bj += 1
            else:
                tmp_P[k] = PA[i, aj]
                tmp_I[k] = IA[i, aj]
                aj += 1
        PA[i] = tmp_P
        IA[i] = tmp_I
```

In my tests, when `PA` (or `PB`) has shape `(50, 100_000)`, the existing function takes around 30 seconds to merge both arrays. However, my simple new function produces the same result but only takes 0.03 seconds!

The new algo is very, very simple. We keep track of two indices `aj` and `bj`, which correspond to a specific column/element in `PA[i]` and `PB[i]`. We start by setting `aj = 0` and `bj =0` (i.e., they are both pointing to the first element in there respective arrays). Also, we have some temporary arrays `tmp_P` and `tmp_I` that are used to store the OVERALL top-k from both arrays. Then, we loop over each element of `tmp_P` and fill it up with the next smallest value either from `PA[i]` or `PB[i]`. 

In the first iteration of this for-loop, we compare the smallest number from either array (i.e., `PA[i, 0]` vs `PB[i, 0]`) and take the smallest number between the two and put it in `tmp_P`. Next, and here's the magic, if we took `PA[i, 0]` then we increment `aj` so that it is no longer pointing to `PA[i, 0]` and, instead, we'll look at `PA[i, 1]`. Similarly, if we took `PB[i, 0]` in the first iteration then we'll increment `bj` so that the smallest value from `PB[i]` that will be referenced will then be `PB[i, 1]`.  So, each time we fill `tmp_P` with a value from either `PA[i]` or `PB[i]` then we only increment the index to that array. This means that `PA[i, aj]` and `PB[i, bj]` are ALWAYS pointing to the smallest (yet-to-be-consumed) value. Luckily, if all `k` elements are filled in `tmp_P` then we can end ahead of time and not need to traverse all elements of either array. This algorithm is very efficient and only needs to look at the essential elements. The trick is incrementing the index to the smallest-so-far value in each array.

Considering the timing that I observed, I'd bet that this will improve your performance values too.

Reply:
This is very interesting and cool! In fact, I recently solved a [LeetCode problem](https://leetcode.com/problems/median-of-two-sorted-arrays/) using a similar concept and I do not know why I couldn't connect that to this one! 

(As I said before, I really like the way you connect the dots and see things! I know....I am not assuming you know all these things and of course there is a will to try AGAIN and maybe do some trial and error as well to make sure everything is completely correct without any bug...  still you are very good at not limiting yourself to an existing solution and you have a strong will to re-think things!!!)

---

This is very cool! We are now avoiding the not-so-cheap `np.searchsorted(...)` operation. Thanks for this suggestion! I will revise this and the pearson version. It  is also easier to read!

----

Do you think the following approach is reasonable for avoiding the allocation of a new memory?

```
@njit
def _merge_topk_PI(PA, PB, IA, IB, tmp_P, tmp_I):
    tmp_P[:, :] = PA    
    tmp_I[:, :] = PB
    
    # rest of algorithm

```

Reply:
@seanlaw 
Also, it would be interesting to run the performance code again and see the numbers! I think we will see some cool results!

Reply:
> This is very cool! We are now avoiding the not-so-cheap np.searchsorted(...) operation. Thanks for this suggestion! I will revise this and the pearson version. It is also easier to read!

Yes! For me, there is certainly "beauty" and performance in code. So, I always strive for something more elegant (which is a combination/balance of "beauty" and performance). In this case, the code was getting pretty ugly and so I just looked up ways to "find top-k elements in two sorted arrays" and borrowed some of the concepts and was surprised to see how well it worked. I was REALLY surprised by the performance but it makes sense.

 > Do you think the following approach is reasonable for avoiding the allocation of a new memory?

No. Considering that the timing is 0.03 seconds, I don't think it's worth it at all. It's too insignificant. That's why I didn't use `parallel=True`. It would've made the code harder to read and we really wouldn't gain much. Right now, the code has a great balance of speed and low additional memory.

Reply:
> I will run them again after revising the code!

Great! 

> you are very good at not limiting yourself to an existing solution and you have a strong will to re-think things

Thank you for recognizing that and thank you for your continued persistence. I know that it can be tiring to iterate SO LONG and to keep trying to "take it to the next level". Sometimes, there's just nothing there and we've wasted our time. I really appreciate it as 99.99% of users will never "know" how much effort goes into writing this code as most will simply call a function and be done. Unfortunately, at the end of the day, "I know" what is going on and it will bother me that I/we haven't given it our best effort. For me, this is research. Iterating and experimenting is at the core of doing research. 😄

Reply:
> No. Considering that the timing is 0.03 seconds, I don't think it's worth it at all. It's too insignificant. That's why I didn't use `parallel=True`

Got it!

Reply:
> Thank you for recognizing that and thank you for your continued persistence. I know that it can be tiring to iterate SO LONG

Not any more! I enjoy this!

> I really appreciate it as 99.99% of users will never "know" how much effort goes into writing this code as most will simply call a function and be done.

I was one of those users :) But, now, I can see all the efforts it take to create something useful AND beautiful!

Reply:
Please post the performance results in the main part of the PR and not in these comments.